### ISSUE TITLE
[go_router] Refreshing a page after login always shows login page

### ISSUE THREAD

**Issue by gagan2001** [flutter/flutter#99239]
I have an initialRoute: '/' , which is the login page. After I login, I am directed to the dashboard. When I refresh the dashboard page, go_router shows first the login page and then dashboard. Not sure, if configuration of routes is wrong or I need to create a splash screen in the initialRoute.

**Comment by theshivamlko** [comment]
I think your code is working you need to maintain Loggedin State in app so doesnt go back to login page. Check your navigation condition after login

**Comment by angelhdzmultimedia** [comment]
You need a Route Guard that checks if isSignedIn is true and redirect to dashboard.

**Comment by gagan2001** [comment]
RouteGuard is there, the point I am trying to make is that when I refresh the page after login, first the login page appears and then the logged in page. It seems the router shows first the '/' screen and then the logged in screen.Expectation is that directly Dashboard should show up on refresh.

**Comment by theshivamlko** [comment]
That's what above both replies says, on refresh you are not checking condition immediately or async that takes some time for your validation. till that the login page shows, make you logic synchronous and checkLogin() should complete before login page opens.
Its a logic issue,

**Comment by danagbemava-nc** [comment]
Hi @gagan2001, please provide a complete minimal reproducible code sample that we can use to verify this. Please also provide the version of go_router you're using as well as the output of `flutter doctor -v`. 

Thank you

**Comment by gagan2001** [comment]
Code for go_router class
```

class GoRouterClass {
  final authState = Auth();

  GoRouterClass(authState);

  late final goRouter = GoRouter(
    initialLocation: '/',
    routes: [
      GoRoute(
          path: '/',
          pageBuilder: (context, state) => NoTransitionPage<void>(
                key: state.pageKey,
                restorationId: state.pageKey.value,
                child: const LoginPage(),
              )),
      GoRoute(
          path: '/search',
          pageBuilder: (context, state) => NoTransitionPage<void>(
              key: state.pageKey,
              restorationId: state.pageKey.value,
              // child: const Search(),
              child: const LoggedPageStructure(
                bodyWidget: SearchCandidateBody(),
              ))),
      GoRoute(
          path: '/queue',
          pageBuilder: (context, state) => NoTransitionPage<void>(
              key: state.pageKey,
              restorationId: state.pageKey.value,
              child: const LoggedPageStructure(
                bodyWidget: QueueBody(),
              )
              )
          ),
      GoRoute(
          path: '/queue/addCandidate',
          pageBuilder: (context, state) => NoTransitionPage<void>(
              key: state.pageKey,
              restorationId: state.pageKey.value,
              child: const LoggedPageStructure(
                bodyWidget: AddCandidate(),
              )
              )
          ),
      GoRoute(
          path: '/register',
          pageBuilder: (context, state) => NoTransitionPage<void>(
                key: state.pageKey,
                restorationId: state.pageKey.value,
                child: const Register(),
              )
          ),
      GoRoute(
          path: '/forgotPassword',
          pageBuilder: (context, state) => NoTransitionPage<void>(
                key: state.pageKey,
                restorationId: state.pageKey.value,
                child: const ForgotPassword(),
              )
          ),
      GoRoute(
          path: '/settings/changePassword',
          pageBuilder: (context, state) => NoTransitionPage<void>(
              key: state.pageKey,
              restorationId: state.pageKey.value,
              // child: const ChangePassword(),
              child: const LoggedPageStructure(
                bodyWidget: ChangePasswordBody(),
              )))
    ],
    errorPageBuilder: (context, state) => MaterialPage<void>(
      key: state.pageKey,
      child: Scaffold(body: Center(child: Text(state.error.toString()))),
    ),

    // redirect to the login page if the user is not logged in
    redirect: (state) {
      // if the user is not logged in, they need to login
      final loggedIn = authState.isLoggedIn;
      final urlLoggingIn = state.subloc == '/';
      final urlRegister = state.subloc == '/register';
      final urlForgotPass = state.subloc == '/forgotPassword';

      if (urlLoggingIn && loggedIn) return '/search';
      if (!urlLoggingIn && !loggedIn && !urlRegister && !urlForgotPass) {
        return '/';
      }

      return null;
    },
    refreshListenable: authState,
  );
}

class LoadingScreen extends StatelessWidget {
  const LoadingScreen({this.from, Key? key}) : super(key: key);
  final String? from;

  @override
  Widget build(BuildContext context) => Scaffold(
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: const [
              CircularProgressIndicator(),
              Text('loading ...'),
            ],
          ),
        ),
      );
}
```
code for main.dart

```void main() {
  GoRouter.setUrlPathStrategy(UrlPathStrategy.path);
  WidgetsFlutterBinding.ensureInitialized();

  runApp(
    MultiProvider(providers: [
      ChangeNotifierProvider<Auth>(
        create: (_) => Auth(),
      ),
      ProxyProvider<Auth, GoRouterClass>(
          update: (_, authObj, __) => GoRouterClass(authObj))
    ], child: const MyApp()),
  );
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Consumer<Auth>(builder: (_, authObj, child) {
      var _router = context.watch<GoRouterClass>().goRouter;
      return MaterialApp.router(
        debugShowCheckedModeBanner: false,
        routeInformationParser: _router.routeInformationParser,
        routerDelegate: _router.routerDelegate
       
      );
    });
  }
}
```
Auth provider
```
class Auth extends ChangeNotifier {
  Auth() {
    checkLoginStatus();
  }

  bool _isLoggedIn = false;
  Exception? _error;
  bool _isLoading = true;

  bool get isLoading => _isLoading;

  bool get isLoggedIn => _isLoggedIn;

  Future<void> logOut() async {
    try {
      _isLoading = true;
      _error = null;
      var _logout = await webLogout(); // get data from API call
      _logout == 'Success' ? _isLoggedIn = false : _isLoggedIn = true;
    } on Exception catch (e) {
      _error = e;
    }
    _isLoading = false;
    notifyListeners();
  }

  Future<void> login(user, pass) async {
    try {
      _isLoading = true;
      _error = null;
      var _login = await webLogin(user, pass); // get data from API call
      _login == 'Success' ? _isLoggedIn = true : _isLoggedIn = false;
    } on Exception catch (e) {
      _error = e;
    }
    _isLoading = false;
    notifyListeners();
  }

  Future<void> checkLoginStatus() async {
    try {
      _isLoading = true;
      _error = null;
      var verify = await verifyLogin(); // get data from API call
      verify == 'Success' ? _isLoggedIn = true : _isLoggedIn = false;
    } on Exception catch (e) {
      _error = e;
    }
    _isLoading = false;
    notifyListeners();
  }
}
```

**Comment by gagan2001** [comment]
> That's what above both replies says, on refresh you are not checking condition immediately or async that takes some time for your validation. till that the login page shows, make you logic synchronous and checkLogin() should complete before login page opens. Its a logic issue,

How do you manage in case you are fetching data from API

**Comment by danagbemava-nc** [comment]
Hi @gagan2001, your sample as it is now contains too many unknowns for me to properly investigate this. Can you provide a reduced test case so I can properly investigate this? You can use dummy widgets instead of your actual widgets. 

Thank you

**Comment by gagan2001** [comment]
@danagbemava-nc i was able to resolve the same by adding more conditions in my redirect: parameter, but, it is very tough to imagine all the state combinations and decide next route

**Comment by danagbemava-nc** [comment]
Considering you have been able to resolve your issue, I'm going to mark this as solved and close this issue.  

If you do run into any more issues do not hesitate to file a new issue. 

Thank you

**Comment by Jairolaya12x** [comment]
@gagan2001 Can u add a example about the fix please, I have the same error, if I refresh the page always show the default page, thanks
